ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java

```diff
@@ -279,6 +292,72 @@ a database ( directory )
     return 0;
   }
 
+  /**
+   * Cleanup/drop tables from the given database which are bootstrapped by input dump dir.
+   * @throws HiveException Failed to drop the tables.
+   * @throws IOException File operations failure.
+   * @throws InvalidInputException Invalid input dump directory.
+   */
+  private void bootstrapRollbackTask() throws HiveException, IOException, InvalidInputException {
+    Path bootstrapDirectory = new PathBuilder(work.bootstrapDumpToRollback)
+            .addDescendant(ReplUtils.INC_BOOTSTRAP_ROOT_DIR_NAME).build();
+    FileSystem fs = bootstrapDirectory.getFileSystem(conf);
+
+    if (!fs.exists(bootstrapDirectory)) {
+      throw new InvalidInputException("Input bootstrap dump directory to rollback doesn't exist: "
+              + bootstrapDirectory);
```

Comment:
Please add a test case covering this error i.e. when an invalid bootstrap dump location is specified. If the specified bootstrap dump (to rollback) location exists, how do we know that it is indeed the bootstrap dump location for external tables and not some other dump location like a genuine incremental dump or a genuine bootstrap dump? We should add testcases for the same as well.

Reply:
This feature is not specific for external tables. The idea is to rollback the tables bootstrapped from given dump irrespective of external or acid or even table level replication.
We expect the input dump to be bootstrap combined in incremental dump. If full bootstrap dump is specified, it throw exception. I will add a test to see if any other dump is specified, then repl load should fail.

Reply:
Done,